noodles-bed/src/ser.rs

```diff
@@ -1,57 +1,62 @@
-use serde::{ser::{self, SerializeStruct}, Serialize};
+use serde::{ser, Serialize};
 
-use crate::{error::{Error, Result}, Record};
+use crate::{
+    error::{Error, Result},
+    record::{AuxiliarBedRecordWrapper, BedN},
+    Record,
+};
 
-pub struct Serializer {
-    // This string starts empty and JSON is appended as values are serialized.
+pub struct Record3Serializer {
     output: String,
```

Comment:
To avoid serializing records manually, you could keep track of a [`bed::record::builder::Builder`](https://docs.rs/noodles-bed/0.4.0/noodles_bed/record/builder/struct.Builder.html) state, and progressively build it up inside the `serialize_*` functions. For example, if a `chromStart` field is expected next, and a `serialize_u64` (or another integer serialize function) is called, then the [`set_start_position`](https://docs.rs/noodles-bed/0.4.0/noodles_bed/record/builder/struct.Builder.html#method.set_start_position) builder function could be called. At the end (or intermittently throughout) the records could be converted to string output. This way all the string parsing is done by the noodles library.

This is similar to keeping track of the state for the `Deserializer`.

Reply:
Actually, we are already using noodles functionality to parse the serialization for us, via the `DisplayFromStr` functionality on `serde_with` used with the `AuxiliarBedRecordWrapper`

```
#[serde_as]
#[derive(Deserialize, Serialize)]
pub struct AuxiliarBedRecordWrapper<T>
where
    T: BedN<3> + std::str::FromStr + fmt::Display,
    <T as std::str::FromStr>::Err: std::fmt::Display,
{
    #[serde_as(as = "DisplayFromStr")]
    pub record: T,
}
```

I think I tipped you in the wrong direction since `impl<'a> ser::Serializer for &'a mut Record3Serializer {}` is really verbose and has lots of implementation that looks like its made to parse bed structs by itself, is that correct?

This implementation could be heavily dried out now that we are using `serde_with`, but we still need some kind of plain serializer to grab thinks between each `\n` and sending directly to `record: T` `Display` method.

Reply:
Ahh yeah, I think I see what you mean. Do you think there is any way to avoid even having to parse the `\n` characters? Ideally there would be no need to touch a raw string, and just allow noodles to do all of this. 

However, it looks like noodles might [write records one by one](https://github.com/zaeleus/noodles/blob/65b77361032557de8e16c94a6e35d84f3ef10f39/noodles-bed/src/writer.rs#L97-L103) anyway, so this might not be possible.

Reply:
It's kind of weird: we need some Serializer or Deserializer specific implementation to exist to have a full pipeline of serializaton and deserialization, but as soon as we arrive on our Serializer, all it needs to do is:

1) Call Display from the specific Record<N> implementation (which is being done by the `serde_with` derive)
2) Insert `\n` on `Vec<T>` serializations.

So it's the most plain implementation of a Serializer trait ever, we probably can and should call `unreachable!()` on most of the functions (I didn't do this yet because maybe serde_with fails.)

Answering your question: I haven't seen a way to write a string separated `\n` inside of noodles, but now that you mention it, maybe implementing a `Display` and `FromStr` for `Vec<Record<N>>` shouldn't be that hard, the weird part is that we would still need an `AuxiliarWrapper` for this `Vec`, and I can't see a way to use the same `AuxiliarWrapper` both for a single and a collection of Records, in the past I've seen this solved by coercing single Records to a collection of one element, but maybe we don't want that.